### ISSUE TITLE
Bug: `no-useless-assignment` unexpected error with increment/decrement

### ISSUE THREAD

**Issue by erosman** [eslint/eslint#19818]
### Environment

Node version: 23.11.0
npm version: 11.3.0
Local ESLint version: 9.28.0
Global ESLint version:
Operating System: Ubuntu 24.04

### What parser are you using?

Default (Espree)

### What did you do?

Unexpected error with increment/decrement

```js
/* eslint no-useless-assignment: "error" */

// shows error
[1, 2, 3].map(i => ++i);
[1, 2, 3].map(i => --i);

// no error
[1, 2, 3].map(i => i + 1);
[1, 2, 3].map(i => i - 1);
```

### What did you expect to happen?

No error

### What actually happened?

> This assigned value is not used in subsequent statements.  ([no-useless-assignment](https://eslint.org/docs/rules/no-useless-assignment))

### Link to Minimal Reproducible Example

[Playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVzZWxlc3MtYXNzaWdubWVudDogXCJlcnJvclwiICovXG5cbi8vIHNob3dzIGVycm9yXG5bMSwgMiwgM10ubWFwKGkgPT4gKytpKTtcblsxLCAyLCAzXS5tYXAoaSA9PiAtLWkpO1xuXG4vLyBubyBlcnJvclxuWzEsIDIsIDNdLm1hcChpID0+IGkgKyAxKTtcblsxLCAyLCAzXS5tYXAoaSA9PiBpIC0gMSk7Iiwib3B0aW9ucyI6eyJydWxlcyI6eyJuby11c2VsZXNzLWFzc2lnbm1lbnQiOlsiZXJyb3IiXX0sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19)

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
I think this works correctly since the updated value of `i` isn't used in subsequent statements, as stated in the error message, so there's no reason to use `++i`/`--i` instead of `i + 1`/`i - 1`.

@eslint/eslint-team thoughts?

**Comment by erosman** [comment]
Please note that the first snippet that shows an error is the shorthand of the second snippet.

```js
// shows error
[1, 2, 3].map(i => ++i);

// no error
[1, 2, 3].map(i => {
  ++i;
  return i;
});
```

**Comment by snitin315** [comment]
The behaviour seems as expected.

> Please note that the first snippet that shows an error is the shorthand of the second snippet.

The rule flags the concise version `(i => ++i)` because it treats the assignment as unused when the assigned variable (`i`) isn’t referenced after the operation. In the block form (`{ ++i; return i; }`), i is explicitly used post-assignment, so it's not flagged.


This isn’t a bug, but the behavior can be confusing; we can treat this as a potential documentation clarification. Appreciate you pointing it out!

**Comment by erosman** [comment]
Please note that the **assigned** variable is **used** nonetheless, which contradicts `no-useless-assignment`.

- In `{ ++i; return i; }`, variable is assigned and used/returned after assignment
- In `(i => ++i)`, variable is assigned and used/returned at the same time

As an example, `[3].map(i => ++i)`, variable `i` is set to `4` and the new assigned value `4` is used, which results in `[4]`.
There is no useless assignment in this example.

**Comment by snitin315** [comment]
You're absolutely right that in `(i => ++i)`, the assigned value is returned and has an observable effect - `[3].map(i => ++i)` indeed results in [4].

That said, the rule is based on the concept of a `dead store`. As described in [Wikipedia](https://en.wikipedia.org/wiki/Dead_store):

> In computer programming, a local variable that is assigned a value but is not read by any subsequent instruction is referred to as a dead store.

In `i => ++i`, while the return value is used, the assigned variable (`i`) itself is not read after the assignment. The rule interprets this as a dead store, the variable is mutated, but never read again, and flags it as potentially unnecessary.

**Comment by erosman** [comment]
It is worth considering the difference between the following examples:

```js
// i is set to 4 and 4 is used
[3].map(i => ++i)
// Array [ 4 ]

// i is set to 4 but 4 is NOT used
[3].map(i => i++)
// Array [ 3 ]
```

**Comment by snitin315** [comment]
Got it, thanks for pointing that out, it makes sense now.

```js
/* eslint no-useless-assignment: "error" */

// These should be OK — the incremented/decremented value is used
[1, 2, 3].map(i => ++i); // [2, 3, 4] OK
[1, 2, 3].map(i => --i); // [0, 1, 2] OK

// These should be reported — the updated value is discarded (dead store)
[1, 2, 3].map(i => i++); // [1, 2, 3] Should report error
[1, 2, 3].map(i => i--); // [1, 2, 3] Should report error
```

The rule should ideally distinguish between prefix (`++i/--i`) and postfix (`i++/i--`) cases, since only the latter results in an actual dead store.

Marking this as accepted.

**Comment by SwetaTanwar** [comment]
I would like to work on the fix for this issue.

**Comment by mdjermanovic** [comment]
> // i is set to 4 and 4 is used
> [3].map(i => ++i)
> // Array [ 4 ]

`++i` expression evaluates to `4`, and that `4` is used.

`i` is set to `4`, but `4` from `i` is never used.

**Comment by mdjermanovic** [comment]
To clarify how `++` works, here's an example with getters and setters:

```js
const obj = {
    get i() {
        // always 3
        return 3;
    },
    set i(value) {
        // do nothing
    } 
}

console.log([obj].map(obj => ++obj.i)); // logs [ 4 ]. The value of `obj.i` is not used after `++obj.i`.

```

**Comment by erosman** [comment]
A **Block-scoped loop** requires specific consideration.

In the following example, `i` in `map()` only exists within the Block-scoped loop. It has no bearing on variables outside the loop.
It is within the Block-scoped loop that it is **set** and **used**.

```js
let i = 20;
[1, 2, 3].map(i => ++i);
console.log(i);
// 20
```

Another example ...

```js
let i = 20;
for (let i = 0; i < 10; i++) {
  console.log(i);
}
console.log(i);
// 20
```

**Comment by ota-meshi** [comment]
> Marking this as accepted.

If you check the spec, GetValue is not called to get the last value, so I think it's a dead store. I don't think it's right for the rule not to report it by default.
It might be possible to make it optional to ignore it.
https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#sec-prefix-increment-operator-runtime-semantics-evaluation

**Comment by mdjermanovic** [comment]
> In the following example, `i` in `map()` only exists within the Block-scoped loop. It has no bearing on variables outside the loop. It is within the Block-scoped loop that it is **set** and **used**.
> 
> ```js
> let i = 20;
> [1, 2, 3].map(i => ++i);
> console.log(i);
> // 20
> ```

The variable is used, but the value assigned to it by `++` is never read from that variable, so this qualifies as a useless assignment.

**Comment by nzakas** [comment]
I believe the rule is working correctly. Consider this example:

```js
let i = 20;
[1, 2, 3].map(j => j+=1);
console.log(i);
```

Here, it's clear that the value of `j` isn't really being used after it's set. You could just as well write this to avoid the error:

```js
let i = 20;
[1, 2, 3].map(j => j+1);
console.log(i);
```

So, I don't think any changes are necessary here aside from update the docs to explain the prefix/postfix increment case more clearly.

**Comment by erosman** [comment]
The example doesn’t show the result of the mapped process. Let's consider:

```js
const arr = [1, 2, 3].map(i => i+=1);
// Array(3) [ 2, 3, 4 ]
```

The process:

- 1 -> `map` as `i`, set `i += 1` return `i` as 2
- 2 -> `map` as `i`, set `i += 1` return `i` as 3
- 3 -> `map` as `i`, set `i += 1` return `i` as 4
- final result `[ 2, 3, 4 ]`

The final value of `i` was set to 4 and `i` was returned to the `arr` as 4.
The final value of `i`, being 4, is used in `arr`.

Nonetheless, the 2 examples in https://github.com/eslint/eslint/issues/19818#issuecomment-2932685467 are **identical** but result in different outcome by ESLint.

**Comment by mdjermanovic** [comment]
> * 3 -> `map` as `i`, set `i += 1` return `i` as 4

It actually works like this:

Step 1: Read value from variable `i`. It's number 3.
Step 2: Calculate 3 + 1 = 4.
Step 3: Set value 4 to `i`.
Step 4: Return value 4.

Value 4 is never read from variable `i`.

> Nonetheless, the 2 examples in [#19818 (comment)](https://github.com/eslint/eslint/issues/19818#issuecomment-2932685467) are **identical** but result in different outcome by ESLint.

In the first example, values that `++` sets to `i` are never read. In the second example, they're read by `return i`.

**Comment by nzakas** [comment]
Okay, I'm in agreement with @mdjermanovic that the rule is behaving as expected. The conversation isn't revealing any new details at this point so I'm closing the issue. Regardless of opinion, this is an edge case of this rule and is best handled using disable comments.
